### DISCUSSION TITLE
How to make GitOps deploy in multiple env's ?

### DISCUSSION THREAD

**Discussion by jurabek** [fluxcd/flux2#715] [Q&A] [answered]
Hi guys, I am really struggling to get an idea for deployment for multiple env's, currently, I am adding the `GitOps+Flux` approach in our company.
The problem is I am bumping the version after merging to master let’s say, I had a version `1.0.0` and  I bumped to `1.1.0` in the manifest file then creating PR for deployment if someone approves the changes, Flux2 will deploy it into ENV’s.
The question is, I have 3 environments `DEV`, `STAGING`, `PROD`, how I will deploy by approving PR’s, should I create 3 PR’s for each env? I would really appreciate any suggestions here.

Thanks,
Jurabek

**Comment by kingdonb** [comment, marked as answer]
Greetings, have you had a look at the https://github.com/fluxcd/flux2-kustomize-helm-example example repo? It shows a solution with two environments, "staging" and "production" – I think this is the recommended approach.

Some background discussion is here: https://github.com/fluxcd/flux/issues/1071 – you may find some helpful information there; the discussion was quite long, and it culminated in the flux2-kustomize-helm-example that I have linked above.

Please have a look at the flux2-kustomize-helm example that was linked above, and let us know if that helps!

**Reply by Edblakejitsuin** [reply]
How do you promote a tested helm chart from staging to production?

**Reply by kingdonb** [reply]
The CI system can have multiple tag and push steps, with differently formed image tags. Ultimately what happens in test and dev environment is more flexible, since you might want to promote every build into a dev env regardless of test status, but only promote passing builds to staging and production.

One strategy that works: use semver tag automation in production. Only tags which have been generated manually or generated by a CI Pass step as semver formatted tags will be promoted into production by a filter for production.

In your lax environment (the dev environment, say) you can promote every build, or every tag generated from a merge into the dev branch.

I think there are more advanced strategies being demonstrated on Flux 2 now, but this strategy has worked well since Flux 1 and continues to work - there is no requirement of communication between CI and CD!

**Reply by o6uoq** [reply]
@kingdonb would you mind sharing the semver tag automation approach? Please and thank you!

**Reply by jurabek** [reply]
@o6uoq recently I implemented `semver tag` automation with `Kustomize + GithubActions` in monorepo, if you want I will share it

**Reply by kingdonb** [reply]
The migration guide has a semver example which is probably the best place to start: https://toolkit.fluxcd.io/guides/flux-v1-automation-migration/#how-to-use-semver-image-tags

Also don't miss this one, https://toolkit.fluxcd.io/guides/flux-v1-automation-migration/#how-to-use-sortable-image-tags

**Reply by o6uoq** [reply]
@Jurabek sure, that would be great! Thank you.

**Reply by jurabek** [reply]
@o6uoq 
I am sharing just POC of my idea that we realized at work.
check it out example [CI/CD pipeline](https://github.com/chayxana/Restaurant-App/blob/develop/.github/workflows/basket-api.yml)

the pipeline contains steps below:
- build (packaging app if passes automatically bumping SemVer and will commit it)
- release (pushing docker image into the registry with new bumped version)
- update manifest files (updates new image tag in `kustomize` and will commit it)

our pipeline commits new "Deployment to commit" automatically into the `develop` branch and then we will create a Pull Request: `from dev -> staging -> master`

manifests folder structure: 
```
├── clusters
│   ├── prod
│   └── staging
├── infrastructure
│   ├── fluentd
│   ├── istio
│   │   ├── istio-init.yaml
│   │   ├── istio-operator.yaml
│   │   └── kustomization.yaml
│   ├── kustomization.yaml
│   └── namespaces.yaml
└── services
    ├── basket.api
    │   ├── base
    │   │   ├── config.env
    │   │   ├── deployment.yaml
    │   │   ├── kustomization.yaml
    │   │   ├── redis.yaml
    │   │   ├── service.yaml
    │   │   └── virtual-service.yaml
    │   ├── prod
    │   │   ├── config.env
    │   │   ├── kustomization.yaml
    │   │   └── vs_hosts_patch.yaml
    │   └── staging
    │       ├── config.env
    │       ├── kustomization.yaml
    │       └── vs_hosts_patch.yaml
    └── order.api
        ├── base
        └── prod
```

don't hesitate if you have any other questions

**Comment by jurabek** [comment]
@kingdonb thanks for your answer, I create a branch per environment and now flux syncing based on branches.

**Comment by g3kk0** [comment]
Just stumbled across this when trying to figure out how to manage progressive rollouts e.g. deploying to a test environment first before rolling out changes to other environments. Using Flux with OCI for this really help. You're able to simply to treat your platform code exactly like your app code and bundle it into a versioned deployable artifact and publish it to an immutable registry (just like you do with Docker). This separation between code and artifact provides real freedom in how and when you want to update things. Also it avoids all the merge hell and entanglement with managing multiple git branches for things.

**Reply by matheuscscp** [reply]
This is what we have for environment promotion at the moment, takes a bit of building but the concept can also be applied to Flux Kustomizations: https://fluxcd.io/flux/use-cases/gh-actions-helm-promotion/
